### PR TITLE
Read(...) endpoint for the resource service

### PR DESCRIPTION
**Pull request by analogue**
### Description

See NET-2687.  I swiped `internal/storage/storage.go` from the latest `boxofrad/storage-backend` branch since it has necessary dependencies.

### Testing & Reproduction steps
`gotestsum --format testname ./agent/grpc-external/services/resource` 

### PR Checklist

* [x] updated test coverage
* [ ] external facing docs updated
* [x] not a security concern

### CODE REVIEW

File: agent/grpc-external/services/resource/read.go
Comment (boxofrad) [inline agent/grpc-external/services/resource/read.go:29]: So sorry, but I've just broken this API in #16538 😅 

It should now be:
```suggestion
	consistency := storage.EventualConsistency
	if isConsistentRead(ctx) {
		consistency = storage.StrongConsistency
	}

	resource, err := s.backend.Read(ctx, consistency, req.Id)
```

File: agent/grpc-external/services/resource/read.go
Comment (boxofrad) [inline agent/grpc-external/services/resource/read.go:17]: Super nit-pick 🐜 would you mind nixing this trailing newline? We don't tend to have them at the beginning of function bodies.

File: agent/grpc-external/services/resource/read_test.go
Comment (boxofrad) [inline agent/grpc-external/services/resource/read_test.go:19]: Is there a reason we can't just use the real registry here? (and in the below test cases too)

File: agent/grpc-external/services/resource/read.go
Comment (boxofrad) [inline agent/grpc-external/services/resource/read.go:1]: Thanks for pulling this out into its own file. That's much better!

**Review by boxofrad — COMMENTED** [review]
This looks great! 👏🏻  As there's still some churn going on in the storage backend interface, let's maybe hold off on merging until #16538 is merged.

File: agent/grpc-external/services/resource/read.go
Comment (analogue) [reply agent/grpc-external/services/resource/read.go]: Fixed

File: agent/grpc-external/services/resource/read.go
Comment (analogue) [reply agent/grpc-external/services/resource/read.go]: 👍

File: agent/grpc-external/services/resource/read_test.go
Comment (analogue) [reply agent/grpc-external/services/resource/read_test.go]: Updated all tests to not use a mock Registry or Backend since `consistency` can now be asserted against.

**Comment by analogue** [comment]
Fixed all outstanding issues.

File: agent/grpc-external/services/resource/read.go
Comment (boxofrad) [inline agent/grpc-external/services/resource/read.go:36]: tip: it's probably worth using [`errors.Is`](https://pkg.go.dev/errors#Is) and [`errors.As`](https://pkg.go.dev/errors#As) here in case backends [wrap the error](https://go.dev/blog/go1.13-errors).

File: agent/grpc-external/services/resource/read_test.go
Comment (boxofrad) [inline agent/grpc-external/services/resource/read_test.go:16]: Lots of folks on the team have their editor set up to split imports into stdlib, project, and 3rd party groups. It's what you get if you run `goimports -local github.com/hashicorp/consul` but I think there's an equivalent [snippet](https://stackoverflow.com/questions/38704769/golang-import-grouping-by-package) for VS Code and GoLand.

I don't think it's a hard rule, but it'll reduce some diff noise for folks that work on this file later. 😄 

```suggestion
import (
	"context"
	"testing"

	"github.com/hashicorp/consul/internal/resource"
	"github.com/hashicorp/consul/internal/storage"
	"github.com/hashicorp/consul/internal/storage/inmem"
	"github.com/hashicorp/consul/proto-public/pbresource"
	"github.com/hashicorp/consul/proto/private/prototest"
	
	"github.com/stretchr/testify/require"
	"google.golang.org/grpc/codes"
	"google.golang.org/grpc/metadata"
	"google.golang.org/grpc/status"
)
```

**Review by boxofrad — COMMENTED** [review]
Awesome work! I've left a couple more small optional comments, but this is good to go as soon as the storage backend is merged 🎉

File: agent/grpc-external/services/resource/read_test.go
Comment (analogue) [reply agent/grpc-external/services/resource/read_test.go]: I ran `goimports` as described and got
```
import (
        context "context"
        "testing"

        mock "github.com/stretchr/testify/mock"
        "github.com/stretchr/testify/require"
        "google.golang.org/grpc/codes"
        "google.golang.org/grpc/metadata"
        "google.golang.org/grpc/status"

        "github.com/hashicorp/consul/internal/resource"
        storage "github.com/hashicorp/consul/internal/storage"
        "github.com/hashicorp/consul/internal/storage/inmem"
        pbresource "github.com/hashicorp/consul/proto-public/pbresource"
        "github.com/hashicorp/consul/proto/private/prototest"
)
```

Ignoring the duplicate aliases, it put `github.com/hashicorp/consul` imports at the end instead of the middle. I don't see any additional flags on `goimports` to control this behavior.

File: agent/grpc-external/services/resource/read.go
Comment (analogue) [reply agent/grpc-external/services/resource/read.go]: TIL! Fixed in https://github.com/hashicorp/consul/pull/16655/commits/d5cd617e19421a2097ed1d404cf497acbb790583

File: agent/grpc-external/services/resource/read_test.go
Comment (analogue) [reply agent/grpc-external/services/resource/read_test.go:104]: Ah! This is leftover from the hybrid use of the non-mock registry and the mock backend. I added a test as suggested in https://github.com/hashicorp/consul/pull/16655/commits/5d8dffb57d961af73a68f90803d0a14267d94187.

File: agent/grpc-external/services/resource/server.go
Comment (boxofrad) [inline agent/grpc-external/services/resource/server.go]: This is really clever btw! I've always struggled to figure out how to generate mocks for interfaces in other packages 😍

**Review by boxofrad — APPROVED** [review]

**Pull request merged**
